File: gospider/extractor/extractor.go
package extractor

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// AIExtractor AI 内容提取器
type AIExtractor struct {
	apiKey     string
	apiURL     string
	model      string
	timeout    time.Duration
	httpClient *http.Client
}

// NewAIExtractor 创建 AI 提取器
func NewAIExtractor(apiKey, apiURL, model string) *AIExtractor {
	return &AIExtractor{
		apiKey:  apiKey,
		apiURL:  apiURL,
		model:   model,
		timeout: 30 * time.Second,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// ExtractResult 提取结果
type ExtractResult struct {
	Data map[string]interface{} `json:"data"`
	Error string `json:"error,omitempty"`
}

// Extract 提取结构化数据
func (e *AIExtractor) Extract(content string, schema map[string]interface{}) (*ExtractResult, error) {
	// 构建请求
	payload := map[string]interface{}{
		"model": e.model,
		"messages": []map[string]string{
			{
				"role": "user",
				"content": "Extract structured data from the following content according to this schema:\n" + content,
			},
		},
	}
	
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	
	req, err := http.NewRequest("POST", e.apiURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}
	
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+e.apiKey)
	
	resp, err := e.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	
	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	
	// 解析响应
	extractResult := &ExtractResult{
		Data: make(map[string]interface{}),
	}
	
	if choices, ok := result["choices"].([]interface{}); ok && len(choices) > 0 {
		if choice, ok := choices[0].(map[string]interface{}); ok {
			if message, ok := choice["message"].(map[string]interface{}); ok {
				if content, ok := message["content"].(string); ok {
					// 尝试解析为 JSON
					json.Unmarshal([]byte(content), &extractResult.Data)
				}
			}
		}
	}
	
	return extractResult, nil
}

// Summarize 总结内容
func (e *AIExtractor) Summarize(content string, maxLength int) (string, error) {
	payload := map[string]interface{}{
		"model": e.model,
		"messages": []map[string]string{
			{
				"role": "user",
				"content": fmt.Sprintf("Summarize the following content in %d characters:\n%s", maxLength, content),
			},
		},
	}
	
	jsonData, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	
	req, err := http.NewRequest("POST", e.apiURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}
	
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+e.apiKey)
	
	resp, err := e.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	
	var result map[string]interface{}
	json.Unmarshal(body, &result)
	
	if choices, ok := result["choices"].([]interface{}); ok && len(choices) > 0 {
		if choice, ok := choices[0].(map[string]interface{}); ok {
			if message, ok := choice["message"].(map[string]interface{}); ok {
				if content, ok := message["content"].(string); ok {
					return content, nil
				}
			}
		}
	}
	
	return "", nil
}

// ExtractKeywords 提取关键词
func (e *AIExtractor) ExtractKeywords(content string, maxKeywords int) ([]string, error) {
	_, err := e.Summarize(content, 500)
	if err != nil {
		return nil, err
	}

	// 简单实现，实际应该调用 AI
	keywords := []string{}
	// 这里可以添加关键词提取逻辑

	return keywords, nil
}

// SetTimeout 设置超时
func (e *AIExtractor) SetTimeout(timeout time.Duration) {
	e.timeout = timeout
	e.httpClient.Timeout = timeout
}

File: pyspider/extractor/extractor.py
"""
AI 内容提取模块
支持 OpenAI/Claude 等 AI API
"""

import json
import requests
from typing import Dict, List, Optional, Any


class AIExtractor:
    """AI 内容提取器"""

    def __init__(
        self,
        api_key: str,
        api_url: str = "https://api.openai.com/v1/chat/completions",
        model: str = "gpt-5.2",
        timeout: int = 30,
    ):
        self.api_key = api_key
        self.api_url = api_url
        self.model = model
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            }
        )

    def extract(
        self,
        content: str,
        schema: Dict[str, Any],
    ) -> Optional[Dict[str, Any]]:
        """提取结构化数据"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Extract structured data from the following content according to this schema:\n{json.dumps(schema)}\n\nContent:\n{content}",
                },
            ],
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                content = result["choices"][0]["message"]["content"]
                return json.loads(content)
        except Exception as e:
            print(f"AI extract error: {e}")

        return None

    def summarize(self, content: str, max_length: int = 200) -> Optional[str]:
        """总结内容"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Summarize the following content in {max_length} characters:\n{content}",
                },
            ],
            "max_tokens": 100,
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                return result["choices"][0]["message"]["content"]
        except Exception as e:
            print(f"AI summarize error: {e}")

        return None

    def extract_keywords(self, content: str, max_keywords: int = 10) -> List[str]:
        """提取关键词"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Extract top {max_keywords} keywords from the following content as a JSON array:\n{content}",
                },
            ],
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                content = result["choices"][0]["message"]["content"]
                return json.loads(content)
        except Exception as e:
            print(f"AI keyword extraction error: {e}")

        return []

    def classify(
        self,
        content: str,
        categories: List[str],
    ) -> Optional[str]:
        """分类内容"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Classify the following content into one of these categories: {', '.join(categories)}\n\nContent:\n{content}\n\nCategory:",
                },
            ],
            "max_tokens": 10,
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                return result["choices"][0]["message"]["content"].strip()
        except Exception as e:
            print(f"AI classification error: {e}")

        return None

    def analyze_sentiment(self, content: str) -> Optional[Dict[str, Any]]:
        """情感分析"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Analyze the sentiment of the following text. Return JSON with 'sentiment' (positive/negative/neutral) and 'confidence' (0-1):\n{content}",
                },
            ],
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                content = result["choices"][0]["message"]["content"]
                return json.loads(content)
        except Exception as e:
            print(f"AI sentiment analysis error: {e}")

        return None

    def translate(self, content: str, target_language: str) -> Optional[str]:
        """翻译"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Translate the following text to {target_language}:\n{content}",
                },
            ],
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                return result["choices"][0]["message"]["content"]
        except Exception as e:
            print(f"AI translation error: {e}")

        return None

    def answer_question(self, context: str, question: str) -> Optional[str]:
        """问答"""
        payload = {
            "model": self.model,
            "messages": [
                {
                    "role": "user",
                    "content": f"Answer the following question based on the provided context:\n\nContext:\n{context}\n\nQuestion: {question}\n\nAnswer:",
                },
            ],
        }

        try:
            resp = self.session.post(self.api_url, json=payload, timeout=self.timeout)
            resp.raise_for_status()

            result = resp.json()
            if "choices" in result and len(result["choices"]) > 0:
                return result["choices"][0]["message"]["content"]
        except Exception as e:
            print(f"AI Q&A error: {e}")

        return None
